contracts/test/integration/L2PlusXNet.t.sol

```diff
@@ -0,0 +1,599 @@
+        self.nodes[lookupId].childActors.add(subnetActor);
+    }
+
+    function registerNewSubnet(
+        SubnetsHierarchy storage self,
+        SubnetID memory parent,
+        address subnetActor,
+        address gateway
+    ) internal returns (SubnetID memory) {
+        // ensure parent exists
+        getSubnetGateway(self, parent);
+
+        SubnetID memory child = SubnetIDHelper.createSubnetId(parent, subnetActor);
+
+        bytes32 lookupId = child.toHash();
+
+        self.actorToLookupKeys[subnetActor] = lookupId;
+        storeNewNode(self, lookupId, child, gateway);
+        linkNewChild(self, lookupId, subnetActor);
+
+        return child;
+    }
+
+    function getGatewayBySubnetActor(
+        SubnetsHierarchy storage self,
+        address subnetActor
+    ) internal view returns (address) {
+        bytes32 lookupId = self.actorToLookupKeys[subnetActor];
+        return getSubnetGateway(self, lookupId);
+    }
+}
+
+interface IGateway {
+    function sendContractXnetMessage(
+        IpcEnvelope memory envelope
+    ) external payable returns (IpcEnvelope memory committed);
+
+    function fund(SubnetID calldata subnetId, FvmAddress calldata to) external payable;
+
+    function fundWithToken(SubnetID calldata subnetId, FvmAddress calldata to, uint256 amount) external;
+}
+
+contract L2PlusSubnetTest is Test, IntegrationTestBase, IIpcHandler {
+    using SubnetIDHelper for SubnetID;
+    using CrossMsgHelper for IpcEnvelope;
+    using GatewayFacetsHelper for GatewayDiamond;
+    using SubnetActorFacetsHelper for SubnetActorDiamond;
+    using AssetHelper for Asset;
+    using FvmAddressHelper for FvmAddress;
+
+    using LibSubnetsHierarchy for SubnetsHierarchy;
+
+    SubnetsHierarchy private subnets;
+    bytes32 private new_topdown_message_topic;
+
+    function setUp() public override {
+        // there seems no auto way to derive the abi in string, check this before running tests, make sure
+        // it's updated with the implementation
+        new_topdown_message_topic = keccak256(
+            "NewTopDownMessage(address,(uint8,((uint64,address[]),(uint8,bytes)),((uint64,address[]),(uint8,bytes)),uint64,uint256,bytes),bytes32)"
+        );
+
+        // get some free money.
+        vm.deal(address(this), 10 ether);
+    }
+
+    // ======= implements ipc handler =======
+
+    /* solhint-disable-next-line unused-vars */
+    function handleIpcMessage(IpcEnvelope calldata envelope) external payable returns (bytes memory ret) {
+        ret = bytes("");
+    }
+
+    receive() external payable {}
+
+    // ======= internal util methods ========
+
+    function executeTopdownMessages(IpcEnvelope[] memory msgs, GatewayDiamond gw) internal {
+        uint256 minted_tokens;
+
+        for (uint256 i; i < msgs.length; ) {
+            minted_tokens += msgs[i].value;
+            unchecked {
+                ++i;
+            }
+        }
```

Comment:
Since tests do not run in a gas restricted environment, we can unclutter this by removing the unchecked increment.